File: app/src/main/java/com/raredev/vcspace/ui/editor/CodeEditorView.java
package com.raredev.vcspace.ui.editor;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.LinearLayout;
import androidx.core.content.res.ResourcesCompat;
import com.raredev.vcspace.databinding.LayoutCodeEditorBinding;
import com.raredev.vcspace.editor.IDECodeEditor;
import com.raredev.vcspace.events.PreferenceChangedEvent;
import com.raredev.vcspace.models.DocumentModel;
import com.raredev.vcspace.task.TaskExecutor;
import com.raredev.vcspace.util.FileUtil;
import com.raredev.vcspace.util.PreferencesUtils;
import com.raredev.vcspace.util.SharedPreferencesKeys;
import io.github.rosemoe.sora.langs.textmate.VCSpaceTMLanguage;
import io.github.rosemoe.sora.langs.textmate.provider.TextMateProvider;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

public class CodeEditorView extends LinearLayout {

  private LayoutCodeEditorBinding binding;
  private DocumentModel document;

  public CodeEditorView(Context context, DocumentModel document) {
    super(context);
    binding = LayoutCodeEditorBinding.inflate(LayoutInflater.from(context));

    setOrientation(VERTICAL);
    removeAllViews();

    addView(
        binding.getRoot(),
        new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT, 1f));

    setDocument(document);
    setLoading(true);
    
    var modified = document.isModified();

    TaskExecutor.executeAsync(
        () -> {
          String content = "";
          if (document.getContent() != null) {
            content = new String(document.getContent());
          } else {
            content = FileUtil.readFile(document.getPath());
          }
          return content;
        },
        (result) -> {
          binding.editor.setText((String) result, null);
          if (!modified) document.markUnmodified();
          configureEditor();
          postRead();
        });
  }

  private void postRead() {
    binding.editor.setEditorLanguage(createLanguage());
    binding.editor.setCursorPosition(document.getPositionLine(), document.getPositionColumn());
    setLoading(false);
  }

  public void reloadEditor() {
    getEditor().setColorScheme(TextMateProvider.getColorScheme());
    getEditor().setEditorLanguage(createLanguage());
  }

  public void saveDocument() {
    if (document.isModified()) {
      FileUtil.writeFile(document.getPath(), getCode());
      document.markUnmodified();
    }
  }

  public String getCode() {
    return binding.editor.getText().toString();
  }

  public void release() {
    binding.editor.release();
  }

  public void undo() {
    if (binding.editor.canUndo()) {
      binding.editor.undo();
    }
  }

  public void redo() {
    if (binding.editor.canRedo()) {
      binding.editor.redo();
    }
  }

  public void setDocument(DocumentModel document) {
    binding.editor.setDocument(document);
    this.document = document;
  }

  public DocumentModel getDocument() {
    return document;
  }

  public IDECodeEditor getEditor() {
    return binding.editor;
  }

  public void setLoading(boolean loading) {
    binding.circularProgressIndicator.setVisibility(loading ? View.VISIBLE : View.GONE);
  }

  public VCSpaceTMLanguage createLanguage() {
    try {
      return TextMateProvider.createLanguage(getEditor(), document.getName());
    } catch (Exception e) {
      return null;
    }
  }

  @Subscribe(threadMode = ThreadMode.MAIN)
  public void onSharedPreferenceChanged(PreferenceChangedEvent event) {
    String key = event.getKey();
    switch (key) {
      case SharedPreferencesKeys.KEY_FONT_SIZE_PREFERENCE:
        updateTextSize();
        break;
      case SharedPreferencesKeys.KEY_EDITOR_TAB_SIZE:
        updateTABSize();
        break;
      case SharedPreferencesKeys.KEY_DELETE_EMPTY_LINE_FAST:
        updateDeleteEmptyLineFast();
        break;
      case SharedPreferencesKeys.KEY_EDITOR_FONT:
        updateEditorFont();
        break;
      case SharedPreferencesKeys.KEY_LINENUMBERS:
        updateLineNumbers();
        break;
    }
  }

  public void configureEditor() {
    updateEditorFont();
    updateTextSize();
    updateTABSize();
    updateLineNumbers();
    updateDeleteEmptyLineFast();

    binding.editor.setInputType(createInputFlags());
  }

  private void updateTextSize() {
    int textSize = PreferencesUtils.getEditorTextSize();
    binding.editor.setTextSize(textSize);
  }

  private void updateTABSize() {
    int tabSize = PreferencesUtils.getEditorTABSize();
    binding.editor.setTabWidth(tabSize);
  }

  private void updateEditorFont() {
    binding.editor.setTypefaceText(
        ResourcesCompat.getFont(getContext(), PreferencesUtils.getSelectedFont()));
    binding.editor.setTypefaceLineNumber(
        ResourcesCompat.getFont(getContext(), PreferencesUtils.getSelectedFont()));
  }

  private void updateDeleteEmptyLineFast() {
    boolean deleteEmptyLineFast = PreferencesUtils.useDeleteEmptyLineFast();
    binding.editor.getProps().deleteEmptyLineFast = deleteEmptyLineFast;
    binding.editor.getProps().deleteMultiSpaces = deleteEmptyLineFast ? -1 : 1;
  }

  private void updateLineNumbers() {
    boolean lineNumbers = PreferencesUtils.lineNumbers();
    binding.editor.setLineNumberEnabled(lineNumbers);
  }

  private void updateNonPrintablePaintingFlags() {
    /*binding.editor.setNonPrintablePaintingFlags(
    CodeEditor.FLAG_DRAW_WHITESPACE_LEADING
        | CodeEditor.FLAG_DRAW_WHITESPACE_INNER
        | CodeEditor.FLAG_DRAW_WHITESPACE_FOR_EMPTY_LINE);*/
  }

  private int createInputFlags() {
    return EditorInfo.TYPE_CLASS_TEXT
        | EditorInfo.TYPE_TEXT_FLAG_MULTI_LINE
        | EditorInfo.TYPE_TEXT_FLAG_NO_SUGGESTIONS
        | EditorInfo.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD;
  }
}

File: app/src/main/java/com/raredev/vcspace/editor/CodeEditorView.kt
package com.raredev.vcspace.editor

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.widget.LinearLayout
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.isVisible
import com.blankj.utilcode.util.FileIOUtils
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.raredev.vcspace.databinding.LayoutCodeEditorBinding
import com.raredev.vcspace.editor.langs.VCSpaceTMLanguage
import com.raredev.vcspace.events.OnPreferenceChangeEvent
import com.raredev.vcspace.extensions.cancelIfActive
import com.raredev.vcspace.providers.GrammarProvider
import com.raredev.vcspace.res.R
import com.raredev.vcspace.utils.PreferencesUtils
import com.raredev.vcspace.utils.SharedPreferencesKeys
import io.github.rosemoe.sora.lang.EmptyLanguage
import io.github.rosemoe.sora.lang.Language
import io.github.rosemoe.sora.langs.textmate.TextMateColorScheme
import io.github.rosemoe.sora.langs.textmate.registry.ThemeRegistry
import io.github.rosemoe.sora.text.LineSeparator
import io.github.rosemoe.sora.widget.schemes.EditorColorScheme
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

@SuppressLint("ViewConstructor")
class CodeEditorView(context: Context, file: File) : LinearLayout(context) {

  private val binding = LayoutCodeEditorBinding.inflate(LayoutInflater.from(context))

  private val editorScope = CoroutineScope(Dispatchers.Default)

  val editor: VCSpaceEditor
    get() = binding.editor

  val modified: Boolean
    get() = editor.modified

  val file: File?
    get() = editor.file

  init {
    EventBus.getDefault().register(this)
    binding.editor.apply {
      colorScheme = createColorScheme()
      lineSeparator = LineSeparator.LF
      this.file = file
    }
    binding.searcher.bindSearcher(editor.searcher)
    addView(binding.root, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
    configureEditor()
    readFile(file)
  }

  private fun readFile(file: File) {
    setLoading(true)
    editorScope.launch {
      val content = FileIOUtils.readFile2String(file)
      val language = createLanguage()

      withContext(Dispatchers.Main) {
        binding.editor.setText(content, null)
        postRead(language)
      }
    }
  }

  private fun postRead(language: Language) {
    setLoading(false)

    editor.setEditorLanguage(language)
  }

  fun confirmReload() {
    if (modified) {
      MaterialAlertDialogBuilder(context)
        .setTitle(R.string.menu_reload)
        .setMessage(R.string.discard_changes)
        .setPositiveButton(R.string.yes) { _, _ -> readFile(file!!) }
        .setNegativeButton(R.string.no, null)
        .show()
    } else readFile(file!!)
  }

  fun undo() = editor.undo()

  fun redo() = editor.redo()

  fun canUndo() = editor.canUndo()

  fun canRedo() = editor.canRedo()

  fun setModified(modified: Boolean) {
    editor.modified = modified
  }

  fun setFile(file: File) {
    editor.file = file
  }

  fun release() {
    EventBus.getDefault().unregister(this)
    editorScope.cancelIfActive("Editor has been released")
    editor.release()
  }

  suspend fun saveFile(): Boolean {
    return if (modified && FileIOUtils.writeFileFromString(file, editor.text.toString())) {
      setModified(false)
      true
    } else false
  }

  fun beginSearchMode() {
    binding.searcher.beginSearchMode()
  }

  @Subscribe(threadMode = ThreadMode.MAIN)
  fun onSharedPreferenceChanged(event: OnPreferenceChangeEvent) {
    when (event.prefKey) {
      SharedPreferencesKeys.KEY_EDITOR_TEXT_SIZE -> updateTextSize()
      SharedPreferencesKeys.KEY_EDITOR_TAB_SIZE -> updateTABSize()
      SharedPreferencesKeys.KEY_STICKY_SCROLL -> updateStickyScroll()
      SharedPreferencesKeys.KEY_FONT_LIGATURES -> updateFontLigatures()
      SharedPreferencesKeys.KEY_WORDWRAP -> updateWordWrap()
      SharedPreferencesKeys.KEY_DELETE_EMPTY_LINE_FAST -> updateDeleteEmptyLineFast()
      SharedPreferencesKeys.KEY_EDITOR_FONT -> updateEditorFont()
      SharedPreferencesKeys.KEY_LINE_NUMBERS -> updateLineNumbers()
      SharedPreferencesKeys.KEY_DELETE_TABS -> updateDeleteTabs()
    }
  }

  private fun configureEditor() {
    updateEditorFont()
    updateTextSize()
    updateTABSize()
    updateStickyScroll()
    updateFontLigatures()
    updateWordWrap()
    updateLineNumbers()
    updateDeleteEmptyLineFast()
    updateDeleteTabs()
  }

  private fun updateTextSize() {
    editor.setTextSize(PreferencesUtils.textSize.toFloat())
  }

  private fun updateTABSize() {
    editor.tabWidth = PreferencesUtils.tabSize
  }

  private fun updateEditorFont() {
    val font = PreferencesUtils.selectedFont
    editor.typefaceText = ResourcesCompat.getFont(context, font)
    editor.typefaceLineNumber = ResourcesCompat.getFont(context, font)
  }

  private fun updateStickyScroll() {
    editor.props.stickyScroll = PreferencesUtils.stickyScroll
  }

  private fun updateFontLigatures() {
    editor.isLigatureEnabled = PreferencesUtils.fontLigatures
  }

  private fun updateWordWrap() {
    editor.isWordwrap = PreferencesUtils.wordWrap
  }

  private fun updateLineNumbers() {
    editor.isLineNumberEnabled = PreferencesUtils.lineNumbers
  }

  private fun updateDeleteEmptyLineFast() {
    editor.props.deleteEmptyLineFast = PreferencesUtils.deleteEmptyLineFast
  }

  private fun updateDeleteTabs() {
    editor.props.deleteMultiSpaces = if (PreferencesUtils.deleteMultiSpaces) -1 else 1
  }

  private fun setLoading(loading: Boolean) {
    binding.progress.isVisible = loading
    editor.isEditable = !loading
  }

  private fun createColorScheme(): EditorColorScheme {
    return try {
      TextMateColorScheme.create(ThemeRegistry.getInstance())
    } catch (e: Exception) {
      EditorColorScheme()
    }
  }

  private suspend fun createLanguage(): Language {
    val scopeName: String? = GrammarProvider.findScopeByFileExtension(file?.extension)

    return if (scopeName != null) {
      VCSpaceTMLanguage.create(scopeName)
    } else EmptyLanguage()
  }
}
